test/integration/controller_instance_test.go

```diff
@@ -0,0 +1,1475 @@
+					Status: v1beta1.ConditionTrue,
+					Reason: "ProvisionedSuccessfully",
+				}
+				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, condition); err != nil {
+					t.Fatalf("error waiting for instance condition: %v", err)
+				}
+			})
+		})
+	}
+}
+
+func TestDeleteServiceInstance(t *testing.T) {
+	key := osb.OperationKey(testOperation)
+
+	cases := []struct {
+		name               string
+		osbResponse        *osb.DeprovisionResponse
+		skipBindingDelete  bool
+		skipInstanceDelete bool
+		testFunction       func(t *controllerTest)
+	}{
+		{
+			name:        "synchronous deprovision",
+			osbResponse: &osb.DeprovisionResponse{},
+		},
+		{
+			name: "asynchronous deprovision with operation key",
+			osbResponse: &osb.DeprovisionResponse{
+				Async:        true,
+				OperationKey: &key,
+			},
+		},
+		{
+			name: "asynchronous deprovision without operation key",
+			osbResponse: &osb.DeprovisionResponse{
+				Async: true,
+			},
+		},
+		{
+			name:               "deprovision instance with binding not deleted",
+			osbResponse:        &osb.DeprovisionResponse{},
+			skipBindingDelete:  true,
+			skipInstanceDelete: true,
+			testFunction: func(ct *controllerTest) {
+				if err := ct.client.ServiceInstances(ct.instance.Namespace).Delete(ct.instance.Name, &metav1.DeleteOptions{}); err != nil {
+					ct.t.Fatalf("instance delete should have been accepted: %v", err)
+				}
+
+				condition := v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionReady,
+					Status: v1beta1.ConditionFalse,
+					Reason: "DeprovisionBlockedByExistingCredentials",
+				}
+				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, condition); err != nil {
+					ct.t.Fatalf("error waiting for instance condition: %v", err)
+				}
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ct := &controllerTest{
+				t:                  t,
+				broker:             getTestBroker(),
+				binding:            getTestBinding(),
+				instance:           getTestInstance(),
+				skipBindingDelete:  tc.skipBindingDelete,
+				skipInstanceDelete: tc.skipInstanceDelete,
+				setup: func(ct *controllerTest) {
+					ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+						Response: tc.osbResponse,
+					}
+				},
+			}
+			ct.run(tc.testFunction)
+		})
+	}
+}
+
+func TestPollServiceInstanceLastOperation(t *testing.T) {
+	cases := []struct {
+		name                         string
+		setupFunc                    func(t *controllerTest)
+		skipVerifyingInstanceSuccess bool
+		skipInstanceDelete           bool
+		verifyCondition              v1beta1.ServiceInstanceCondition
+		preDeleteBroker              func(t *controllerTest)
+		preCreateInstance            func(t *controllerTest)
+		postCreateInstance           func(t *controllerTest)
+	}{
+		{
+			name: "provisioning with last operation response state in progress",
```

Comment:
I'm not really sure what to check for then. Should this entire test just be deleted?

Reply:
I recommend that you pull `TestAsyncProvisionWithMultiplePolls` into controller_instance_test.go and expand it to include a number of scenarios around provisioning instances asynchronously. This will test in a more complete scenario the same things that are being tested in `TestPollServiceInstanceLastOperation`. For example, it would be good to have at a minimum the following scenarios.
1) Async provision gets a success response on first poll. This may already be covered by another test, but won't hurt to have it here with these as well.
2) Async provision polls a few times before completing successfully. This is what is currently being done in `TestAsyncProvisionWithMultiplePolls`.
3) Async provision gets a failure response on first poll.
4) Async provision polls a few times before the broker returns with a failure.
5) Async provision gets various errors on first poll.
6) Async provision polls a few times before getting various errors.
7) Async provision polls a few times before the user deletes the ServiceInstance. This may be better off in a different test as it touches on deletion behavior.

Reply:
Your current test is covering most of the scenarios where the success/failure/error happens on the first poll. We just need to expand it to make sure that the same behavior happens when it is not the first poll. And a test should, generally speaking, not end in the middle of a provision, as that is not a complete scenario.